worker/rhpv3.go

```diff
@@ -294,9 +398,88 @@ func (a *accounts) deriveAccountKey(hostKey types.PublicKey) types.PrivateKey {
 	return pk
 }
 
+func (r *hostV3) Contract() types.FileContractID {
+	return r.fcid
+}
+
+func (r *hostV3) HostKey() types.PublicKey {
+	return r.acc.host
+}
+
+func (*hostV3) UploadSector(ctx context.Context, sector *[rhpv2.SectorSize]byte) (types.Hash256, error) {
+	panic("not implemented")
+}
+
+func (*hostV3) DeleteSectors(ctx context.Context, roots []types.Hash256) error {
+	panic("not implemented")
+}
+
+func (r *hostV3) DownloadSector(ctx context.Context, w io.Writer, root types.Hash256, offset, length uint64) (err error) {
+	// return errGougingHost if gouging checks fail
+	if errs := PerformGougingChecks(ctx, nil, r.pt).CanDownload(); len(errs) > 0 {
+		return fmt.Errorf("failed to download sector, %w: %v", errGougingHost, errs)
+	}
+	// return errBalanceInsufficient if balance insufficient
+	defer func() {
+		if isBalanceInsufficient(err) {
+			err = fmt.Errorf("%w, err: %v", errBalanceInsufficient, err)
+		}
+	}()
+
+	return r.acc.WithWithdrawal(ctx, func() (amount types.Currency, err error) {
+		err = withTransportV3(ctx, r.siamuxAddr, r.HostKey(), func(t *rhpv3.Transport) error {
+			cost, err := readSectorCost(r.pt)
+			if err != nil {
+				return err
+			}
+
+			var refund types.Currency
+			payment := rhpv3.PayByEphemeralAccount(r.acc.id, cost, r.bh+defaultWithdrawalExpiryBlocks, r.sk)
+			cost, refund, err = RPCReadSector(t, w, r.pt, &payment, offset, length, root, true)
+			amount = cost.Sub(refund)
+			return err
+		})
+		return
+	})
+}
+
+// readSectorCost returns an overestimate for the cost of reading a sector from a host
+func readSectorCost(pt *rhpv3.HostPriceTable) (types.Currency, error) {
+	cost, overflow := pt.InitBaseCost.AddWithOverflow(pt.ReadBaseCost)
+	if overflow {
+		return types.ZeroCurrency, errors.New("overflow occurred while calculating read sector cost, base cost overflow")
+	}
+
+	ulbw, overflow := pt.UploadBandwidthCost.Mul64WithOverflow(modules.SectorSize)
+	if overflow {
+		return types.ZeroCurrency, errors.New("overflow occurred while calculating read sector cost, upload bandwidth overflow")
+	}
+
+	dlbw, overflow := pt.DownloadBandwidthCost.Mul64WithOverflow(modules.SectorSize)
+	if overflow {
+		return types.ZeroCurrency, errors.New("overflow occurred while calculating read sector cost, download bandwidth overflow")
+	}
+
+	bw, overflow := ulbw.AddWithOverflow(dlbw)
+	if overflow {
+		return types.ZeroCurrency, errors.New("overflow occurred while calculating read sector cost, bandwidth overflow")
+	}
+
+	cost, overflow = cost.AddWithOverflow(bw)
+	if overflow {
+		return types.ZeroCurrency, errors.New("overflow occurred while calculating read sector cost")
+	}
+
+	cost, overflow = cost.Mul64WithOverflow(2) // overestimate the cost
+	if overflow {
+		return types.ZeroCurrency, errors.New("overflow occurred while calculating read sector cost, estimate overflow")
+	}
+	return cost, nil
```

Comment:
I think I indeed missed this in the first review. I don't think we need to overestimate a lot here. The only thing we can't accurately determine is bandwidth but that shouldn't be a lot more than a sector worth of data. 

So what I would do is use the accurate numbers then add 16 kib or so of leeway for upload and download bandwidth and try it in production. If it works, it should always work.